app/api/src/sonicpi_api.cpp

```diff
@@ -0,0 +1,706 @@
+        {
+            LOG(ERR, "!Unrecognized Port - " << port.first);
+            missingPort = true;
+            continue;
+        }
+
+        auto portNum = std::stoi(port.second);
+        m_ports[portId] = portNum;
+
+        // Check we can open the port
+        oscpkt::UdpSocket sock;
+        sock.bindTo(portNum);
+        if ((portNum < 1024) || (!sock.isOk()))
+        {
+            LOG(ERR, port.first << ":" << portNum << " [Not Available] ");
+            missingPort = true;
+        }
+        else
+        {
+            LOG(INFO, port.first << ":" << portNum << " [OK] ");
+        }
+        sock.close();
+    }
+
+    if (missingPort)
+    {
+        LOG(ERR, "Critical Error. One or more ports is not available.");
+        m_pClient->ReportStartupError("One or more ports is not available. Is Sonic Pi already running? If not, please reboot your machine and try again.");
+        return false;
+    }
+    else
+    {
+        LOG(INFO, "All Ports OK!");
+    }
+
+    return true;
+}
+
+bool SonicPiAPI::StartRubyServer()
+{
+    // kill any zombie processes that may exist
+    // better: test to see if UDP ports are in use, only kill/sleep if so
+    // best: kill SCSynth directly if needed
+    std::vector<std::string> args;
+    args.push_back(GetPath(SonicPiPath::RubyPath).string());
+    args.push_back("--enable-frozen-string-literal");
+    args.push_back("-E");
+    args.push_back("utf-8");
+    args.push_back(GetPath(SonicPiPath::RubyServerPath).string());
+
+    if (m_protocol == APIProtocol::TCP)
+    {
+        args.push_back("-t");
+    }
+    else
+    {
+        args.push_back("-u");
+    }
+
+    args.push_back(std::to_string(GetPort(SonicPiPortId::server_listen_to_gui)));
+    args.push_back(std::to_string(GetPort(SonicPiPortId::server_send_to_gui)));
+    args.push_back(std::to_string(GetPort(SonicPiPortId::scsynth)));
+    args.push_back(std::to_string(GetPort(SonicPiPortId::scsynth_send)));
+    args.push_back(std::to_string(GetPort(SonicPiPortId::server_osc_cues)));
+    args.push_back(std::to_string(GetPort(SonicPiPortId::erlang_router)));
+    args.push_back(std::to_string(GetPort(SonicPiPortId::websocket)));
+
+    LOG(INFO, "Launching Sonic Pi Runtime Server:");
+
+    std::ostringstream str;
+    for (auto& arg : args)
+    {
+        str << arg << " ";
+    }
+    LOG(INFO, "Args: " << str.str());
+
+    if (m_homeDirWriteable && (m_logOption != LogOption::Console))
+    {
+        m_spRubyServer = StartProcess(args, GetPath(SonicPiPath::ServerOutputLogPath).string(), GetPath(SonicPiPath::ServerErrorLogPath).string());
+    }
+    else
+    {
+        m_spRubyServer = StartProcess(args);
+    }
+
+    if (!m_spRubyServer)
+    {
+        m_pClient->ReportStartupError("The Sonic Pi Server could not be started!");
+        LOG(ERR, "Failed to start ruby server!");
+        return false;
+    }
+
+#if WIN32
+    // TODO: This fails to find wmic.  And probably needs special UAC support to raise privalige
```

Comment:
*Privilege

Reply:
Thanks for catching these; I'm fixing as I go, and as I refactor stuff.

Reply:
Not quite sure what you mean - do I wait until you've finished the PR then?